src/stdweb.rs

```diff
@@ -1545,11 +1533,9 @@ impl HasContext for Context {
     }
 
     unsafe fn uniform_1_i32(&self, uniform_location: Option<Self::UniformLocation>, x: i32) {
-        let uniform_locations = self.uniform_locations.borrow();
-        let raw_uniform_location = uniform_location.map(|u| uniform_locations.1.get_unchecked(u));
         match self.raw {
-            RawRenderingContext::WebGL1(ref gl) => gl.uniform1i(raw_uniform_location, x),
-            RawRenderingContext::WebGL2(ref gl) => gl.uniform1i(raw_uniform_location, x),
+            RawRenderingContext::WebGL1(ref gl) => gl.uniform1i(uniform_location.as_ref(), x),
```

Comment:
Won't this require us to clone `Self::UniformLocation` whenever we want to use these functions multiple times on the same uniform location? I was thinking we could change the signatures to accept `Option<&Self::UniformLocation>` instead, so the caller would be responsible for borrowing it. What do you think?

Reply:
You are right. But I'm not sure changing the signatures to `Option<&Self::UniformLocation>` would be much better, because the caller would also have to do different things depending on the backend used. On native, `.as_ref()` wouldn't be needed in the caller side, but on web it would be required.

I'm guessing the only good solution would be to implement a new type containing WebGlUniformLocation and implementing Copy. That way, code on both platforms would work seamlessly in all cases. Does it sound reasonable?

Reply:
Another option is to keep `Copy` removed for now and have a new type (which is `Clone` but not `Copy`) containing the native type. So then both backends would use references or `Clone` whenever necessary

Reply:
Correct me if I'm wrong. I think there are basically 3 options:

1. Let it like the PR is now. So the associated type only has `Clone + Debug`, and the user would have to do `.clone()` when she needs it. Good for the crowd of explicit better than implicit, but maybe a bit overkill in that sense.

2. Add `Copy` to the associated type, and implement a new container type around the WebGLUniformLocation. That way, no need to do `.clone()` anywhere. But since the implementation of the opaque type WebGlUniformLocation is not guaranteed to be trivially copyable, it could theoretically bite us in the future (in practice, I don't think so).

3. Remove the associated type entirely, and expose the UniformLocation type with `#[cfg(...)]` trickery. This would allow to use `.as_ref()` in caller site, and change all the fn signatures accordingly as you proposed in the first message. Probably the least abstract and opinionated solution, but opens the possibility of something compiling in one target but not in other.

Reply:
Yeah I think that's right. Option 1 sounds best to me :+1:

Reply:
Yeah, I think so too.